In [8]:
!pip install huggingface_hub

In [2]:
import pandas as pd
import numpy as np

# Define the special Hugging Face URL for the CSV file
# This tells pandas to use fsspec and huggingface_hub to find and read the file
hf_csv_url = "hf://datasets/MatrixIA/FraudData/FraudData.csv"

print(f"Attempting to load full dataset from: {hf_csv_url}")
try:
    # Use pandas read_csv directly with the hf:// URL
    # This loads the entire dataset into memory. May take a minute or two.
    df_full = pd.read_csv(hf_csv_url)
    print("Successfully loaded full dataset into Pandas DataFrame.")
    print(f"Full DataFrame shape: {df_full.shape}")

    # --- IMPORTANT: Create a Sample for Development ---
    # Define sample size (e.g., 1 million rows)
    sample_size = 1000000
    print(f"Creating a sample of {sample_size} rows for initial development...")

    # Option 1: Take the first N rows (simplest)
    df = df_full.head(sample_size).copy()

    # Option 2: Take a random sample (better representation, might be slightly slower)
    # df = df_full.sample(n=sample_size, random_state=42).copy()

    print(f"Sample DataFrame shape: {df.shape}")

    # Optional: Delete the full dataframe to free memory if you notice slowdowns
    # Although with 24GB RAM, it might not be necessary yet.
    # del df_full
    # import gc # Garbage collector
    # gc.collect() # Force memory cleanup

except Exception as e:
    # Catch potential errors during download or reading
    print(f"ERROR: Failed to load dataset using pd.read_csv('hf://...'). Error: {e}")
    print("Check your internet connection, proxy settings (if any), and the URL.")
    raise # Stop execution

# --- Basic Inspection (Run on the SAMPLE 'df') ---
# This part remains the same as before, using the 'df' variable which now holds the sample
print("\nFirst 5 rows (from sample):")
print(df.head())

print("\nDataFrame Info (from sample):")
df.info()

print("\nSummary Statistics (from sample):")
print(df.describe())

print("\nMissing Values Count per Column (from sample):")
print(df.isnull().sum())

# --- Target Variable Check (Run on the SAMPLE 'df') ---
# Target column is 'isFraud' for this dataset (PaySim)
target_column = 'isFraud'

if target_column in df.columns:
    print(f"\nClass Distribution for Target Column ('{target_column}') in Sample:")
    print(df[target_column].value_counts(normalize=True))
    print("\nRaw Counts in Sample:")
    print(df[target_column].value_counts(normalize=False))
else:
    print(f"\nERROR: Target column '{target_column}' not found in the DataFrame!")
    print(f"Available columns are: {list(df.columns)}")

Attempting to load full dataset from: hf://datasets/MatrixIA/FraudData/FraudData.csv


'(ReadTimeoutError("HTTPSConnectionPool(host='huggingface.co', port=443): Read timed out. (read timeout=10)"), '(Request ID: 9331ebc1-4021-401a-b701-138764da7a4f)')' thrown while requesting GET https://huggingface.co/datasets/MatrixIA/FraudData/resolve/main/FraudData.csv
Retrying in 1s [Retry 1/5].


Successfully loaded full dataset into Pandas DataFrame.
Full DataFrame shape: (6362620, 11)
Creating a sample of 1000000 rows for initial development...
Sample DataFrame shape: (1000000, 11)

First 5 rows (from sample):
   step      type    amount     nameOrig  oldbalanceOrg  newbalanceOrig  \
0     1   PAYMENT   9839.64  C1231006815       170136.0       160296.36   
1     1   PAYMENT   1864.28  C1666544295        21249.0        19384.72   
2     1  TRANSFER    181.00  C1305486145          181.0            0.00   
3     1  CASH_OUT    181.00   C840083671          181.0            0.00   
4     1   PAYMENT  11668.14  C2048537720        41554.0        29885.86   

      nameDest  oldbalanceDest  newbalanceDest  isFraud  isFlaggedFraud  
0  M1979787155             0.0             0.0        0               0  
1  M2044282225             0.0             0.0        0               0  
2   C553264065             0.0             0.0        1               0  
3    C38997010         21182.0   

In [3]:
# Import the profiling tool
try:
    from ydata_profiling import ProfileReport
except ImportError:
    try:
        from pandas_profiling import ProfileReport
    except ImportError:
        print("ERROR: Neither ydata-profiling nor pandas-profiling seem to be installed.")
        print("Please run: pip install ydata-profiling")
        raise # Stop if library isn't installed

print("\nGenerating data profiling report on the SAMPLE... (This might take a minute or two)")

# Create the report object using the SAMPLE DataFrame 'df'
profile = ProfileReport(df, title="Fraud Data Profiling Report (Sample)", explorative=True)

# Define the filename for the HTML report
report_filename = "fraud_data_profiling_report_sample.html"

# Save the report to an HTML file
profile.to_file(report_filename)

print(f"Profiling report saved to {report_filename}")


Generating data profiling report on the SAMPLE... (This might take a minute or two)


Summarize dataset:   0%|          | 0/5 [00:00<?, ?it/s]


100%|████████████████████████████████████████████████████████████████| 11/11 [00:09<00:00,  1.14it/s]


Generate report structure:   0%|          | 0/1 [00:00<?, ?it/s]

Render HTML:   0%|          | 0/1 [00:00<?, ?it/s]

Export report to file:   0%|          | 0/1 [00:00<?, ?it/s]

Profiling report saved to fraud_data_profiling_report_sample.html


In [4]:
#Step 2 Level 1: Preprocessing-and-Baseline
# --- Step 2: Basic Preprocessing ---
print("--- Starting Step 2: Basic Preprocessing ---")

# --- Step 2.1: Feature Selection ---
# First, ensure 'df' holds the sample DataFrame from Step 1
# (If in a new notebook, you might need to re-load/re-sample or load from a saved file)
print(f"Original sample shape: {df.shape}")

# Define the features we decided to keep
features_to_keep = ['step', 'type', 'amount', 'oldbalanceOrg', 'newbalanceOrig', 'oldbalanceDest', 'newbalanceDest']
target = 'isFraud' # Define the target column name

print(f"Keeping features: {features_to_keep}")
print(f"Target variable: {target}")

try:
    # Create a new DataFrame containing only the selected features and the target
    # Using .copy() prevents accidental changes to the original 'df'
    df_processed = df[features_to_keep + [target]].copy()

    print("\nCreated df_processed with selected columns.")
    print(f"Shape after selecting features: {df_processed.shape}")
    print("First 5 rows of df_processed:")
    print(df_processed.head())
except KeyError as e:
    print(f"ERROR: A specified column was not found in the DataFrame: {e}")
    print("Please check the 'features_to_keep' list and the 'target' variable against the columns from df.info().")
    raise # Stop execution if columns are incorrect
except Exception as e:
    print(f"An unexpected error occurred during feature selection: {e}")
    raise

--- Starting Step 2: Basic Preprocessing ---
Original sample shape: (1000000, 11)
Keeping features: ['step', 'type', 'amount', 'oldbalanceOrg', 'newbalanceOrig', 'oldbalanceDest', 'newbalanceDest']
Target variable: isFraud

Created df_processed with selected columns.
Shape after selecting features: (1000000, 8)
First 5 rows of df_processed:
   step      type    amount  oldbalanceOrg  newbalanceOrig  oldbalanceDest  \
0     1   PAYMENT   9839.64       170136.0       160296.36             0.0   
1     1   PAYMENT   1864.28        21249.0        19384.72             0.0   
2     1  TRANSFER    181.00          181.0            0.00             0.0   
3     1  CASH_OUT    181.00          181.0            0.00         21182.0   
4     1   PAYMENT  11668.14        41554.0        29885.86             0.0   

   newbalanceDest  isFraud  
0             0.0        0  
1             0.0        0  
2             0.0        1  
3             0.0        1  
4             0.0        0  


In [5]:
# --- Step 2.2: Encode Categorical Features ('type') ---

# Identify the categorical column(s) remaining in df_processed
categorical_cols = ['type']

# Check current data types and unique values in 'type' before encoding
print("\nData types before encoding:")
print(df_processed.info()) # Check df_processed specifically
if 'type' in df_processed.columns:
    print(f"\nUnique values in 'type' column:\n{df_processed['type'].unique()}")
    print(f"\nValue counts for 'type':\n{df_processed['type'].value_counts()}")
else:
    print("\n'type' column not found in df_processed (already dropped or renamed?).")


print(f"\nApplying one-hot encoding to: {categorical_cols} using pd.get_dummies...")

try:
    # Use pd.get_dummies to convert the 'type' column
    # drop_first=True removes redundancy (prevents multicollinearity)
    df_processed = pd.get_dummies(df_processed, columns=categorical_cols, drop_first=True) # This line reassigns df_processed

    print("\nDataFrame after one-hot encoding:")
    print(df_processed.head()) # Notice the original 'type' column is gone
                               # and new 'type_TRANSFER', 'type_PAYMENT', etc. columns appear
    print(f"\nShape after encoding: {df_processed.shape}")

    # Verify that all columns (except potentially target) are now numerical
    print("\nData types after encoding:")
    df_processed.info()

except KeyError as e:
    print(f"ERROR: Column specified for encoding not found: {e}")
    raise
except Exception as e:
    print(f"An unexpected error occurred during one-hot encoding: {e}")
    raise


Data types before encoding:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000000 entries, 0 to 999999
Data columns (total 8 columns):
 #   Column          Non-Null Count    Dtype  
---  ------          --------------    -----  
 0   step            1000000 non-null  int64  
 1   type            1000000 non-null  object 
 2   amount          1000000 non-null  float64
 3   oldbalanceOrg   1000000 non-null  float64
 4   newbalanceOrig  1000000 non-null  float64
 5   oldbalanceDest  1000000 non-null  float64
 6   newbalanceDest  1000000 non-null  float64
 7   isFraud         1000000 non-null  int64  
dtypes: float64(5), int64(2), object(1)
memory usage: 61.0+ MB
None

Unique values in 'type' column:
['PAYMENT' 'TRANSFER' 'CASH_OUT' 'DEBIT' 'CASH_IN']

Value counts for 'type':
type
CASH_OUT    362676
PAYMENT     329753
CASH_IN     218673
TRANSFER     82424
DEBIT         6474
Name: count, dtype: int64

Applying one-hot encoding to: ['type'] using pd.get_dummies...

DataFrame after one-

In [6]:
# --- Step 1.2.3: Separate Features (X) and Target (y) ---

# Ensure the target variable name is correctly defined
target = 'isFraud'

print("\n--- Step 1.2.3: Separate Features (X) and Target (y) ---")
print(f"\nSeparating features (X) and target ('{target}')...")

try:
    # IMPORTANT: Make sure 'df_processed' is the DataFrame from the previous step
    # containing the one-hot encoded 'type' columns.

    # Create the features DataFrame 'X' by dropping the target column
    # axis=1 specifies we are dropping a column
    X = df_processed.drop(target, axis=1) # X should have 10 columns

    # Create the target Series 'y' by selecting only the target column
    y = df_processed[target] # y should be a Series

    # --- Verification ---
    print("Separation complete.")
    print("\nFeatures (X) verification:")
    print(f"  Shape: {X.shape}") # Expect (1000000, 10)
    if target in X.columns:
         print(f"  ERROR: Target column '{target}' still present in X!")
    else:
         print(f"  Target column '{target}' successfully removed from X.")
    print("  First 5 rows of X:")
    print(X.head())
    print("\nTarget (y) verification:")
    print(f"  Shape: {y.shape}") # Expect (1000000,)
    print(f"  Data type: {y.dtype}") # Expect int64
    print("  First 5 values of y:")
    print(y.head())

except KeyError:
    print(f"ERROR: Could not find target column '{target}' in df_processed.")
    print(f"Available columns are: {list(df_processed.columns)}")
    raise
except Exception as e:
    print(f"An unexpected error occurred during X/y separation: {e}")
    raise


--- Step 1.2.3: Separate Features (X) and Target (y) ---

Separating features (X) and target ('isFraud')...
Separation complete.

Features (X) verification:
  Shape: (1000000, 10)
  Target column 'isFraud' successfully removed from X.
  First 5 rows of X:
   step    amount  oldbalanceOrg  newbalanceOrig  oldbalanceDest  \
0     1   9839.64       170136.0       160296.36             0.0   
1     1   1864.28        21249.0        19384.72             0.0   
2     1    181.00          181.0            0.00             0.0   
3     1    181.00          181.0            0.00         21182.0   
4     1  11668.14        41554.0        29885.86             0.0   

   newbalanceDest  type_CASH_OUT  type_DEBIT  type_PAYMENT  type_TRANSFER  
0             0.0          False       False          True          False  
1             0.0          False       False          True          False  
2             0.0          False       False         False           True  
3             0.0           Tr

Excellent! That output is perfect and confirms you have successfully completed Step 2.3: Separate Features (X) and Target (y).

X now has the correct shape (1M rows, 10 feature columns).

y now has the correct shape (1M rows, 1 target column as a Series) and data type (int64).

The verification checks confirm the target was correctly removed from X.

In [7]:
#Sub-step 1.3 - Train/Test Split
from sklearn.model_selection import train_test_split
print("\n--- Step 1.3: Train/Test Split ---")


# Ensure X and y exist and are not empty
if not 'X' in locals() or X.empty or not 'y' in locals() or y.empty:
     print("ERROR: Features (X) or target (y) not defined or empty.")
     raise NameError("X or y not defined/empty.")


try:
    X_train, X_test, y_train, y_test = train_test_split(
        X, y,
        test_size=0.2,
        random_state=50,
        stratify=y # Preserve class distribution
    )
    print("Data split into training and testing sets.")
    print(f"X_train shape: {X_train.shape}, y_train shape: {y_train.shape}")
    print(f"X_test shape: {X_test.shape}, y_test shape: {y_test.shape}")
    print("\nTraining set 'isFraud' distribution:")
    print(y_train.value_counts(normalize=True))
    print("\nTest set 'isFraud' distribution:")
    print(y_test.value_counts(normalize=True))
except Exception as e:
    print(f"ERROR during train/test split: {e}")
    raise


--- Step 1.3: Train/Test Split ---
Data split into training and testing sets.
X_train shape: (800000, 10), y_train shape: (800000,)
X_test shape: (200000, 10), y_test shape: (200000,)

Training set 'isFraud' distribution:
isFraud
0    0.999465
1    0.000535
Name: proportion, dtype: float64

Test set 'isFraud' distribution:
isFraud
0    0.999465
1    0.000535
Name: proportion, dtype: float64


In [8]:
!pip install imblearn

In [9]:
#Sub-Step 1.4: Handle Class Imbalance (SMOTE)
from imblearn.over_sampling import SMOTE # Make sure imblearn is installed
print("\n--- Step 1.4: Apply SMOTE to Training Data ---")


# Ensure X_train and y_train exist
if not 'X_train' in locals() or not 'y_train' in locals():
     print("ERROR: X_train or y_train not defined.")
     raise NameError("X_train/y_train missing.")


try:
    # Instantiate SMOTE - random_state for reproducibility, n_jobs to speed up if possible
    smote = SMOTE(random_state=50)
    print(f"Original training shape: X={X_train.shape}, y={y_train.shape}")
    print(f"Original training class distribution:\n{y_train.value_counts(normalize=True)}")


    print("\nApplying SMOTE (this might take a moment)...")
    # Fit SMOTE and resample ONLY the training data
    X_train_resampled, y_train_resampled = smote.fit_resample(X_train, y_train)


    print("SMOTE application complete.")
    print(f"Resampled training shape: X={X_train_resampled.shape}, y={y_train_resampled.shape}")
    print(f"\nResampled training class distribution:\n{y_train_resampled.value_counts(normalize=True)}") # Should be balanced
except Exception as e:
    print(f"ERROR during SMOTE: {e}")
    raise


--- Step 1.4: Apply SMOTE to Training Data ---
Original training shape: X=(800000, 10), y=(800000,)
Original training class distribution:
isFraud
0    0.999465
1    0.000535
Name: proportion, dtype: float64

Applying SMOTE (this might take a moment)...
SMOTE application complete.
Resampled training shape: X=(1599144, 10), y=(1599144,)

Resampled training class distribution:
isFraud
0    0.5
1    0.5
Name: proportion, dtype: float64


EXPLANATION ON 1.4
Original training shape: X=(800000, 10), y=(800000,)
This shows the size of your training dataset before applying SMOTE.
You had 800,000 samples (rows).
X_train had 10 feature columns.
y_train had 800,000 corresponding labels.
Original training class distribution:
isFraud
0 0.999465: Roughly 99.95% of your original training data was labelled 'Not Fraud' (class 0).
1 0.000535: Only about 0.05% of your original training data was labelled 'Fraud' (class 1).
This confirms the severe class imbalance that SMOTE is designed to address. The model would likely ignore the tiny 'Fraud' class if trained on this original data.
Applying SMOTE (this might take a moment)...
This indicates that the smote.fit_resample(X_train, y_train) command started executing. Since n_jobs was removed, it likely ran on a single CPU core.
SMOTE application complete.
Confirmation that the process finished without errors.
Resampled training shape: X=(1599144, 10), y=(1599144,)
This shows the size of your training dataset after applying SMOTE (X_train_resampled, y_train_resampled).
The number of samples has significantly increased to 1,599,144.
Why the increase? SMOTE works by oversampling the minority class. It doesn't remove majority samples. It creates new, synthetic samples for the minority ('Fraud') class until the number of minority samples equals the number of majority samples.
The number of features (10) remains the same, as SMOTE only creates new samples (rows), not new features.
Resampled training class distribution:
isFraud
0 0.5: Exactly 50% of the resampled training data is now 'Not Fraud'.
1 0.5: Exactly 50% of the resampled training data is now 'Fraud'.
This confirms the successful outcome of SMOTE: the training dataset is now perfectly balanced.
In essence: SMOTE successfully addressed the class imbalance by generating synthetic 'Fraud' examples, resulting in a larger, balanced training dataset (X_train_resampled, y_train_resampled). This balanced dataset will now be used to train the XGBoost model in the next step, forcing the model to pay equal attention to both 'Fraud' and 'Not Fraud' patterns, which is crucial for achieving good Recall on the minority class.

In [10]:
#Sub-Step 1.5: Train Baseline Model (XGBoost)
import xgboost as xgb
print("\n--- Step 1.5: Train Baseline XGBoost Model ---")


# Ensure resampled training data exists
if not 'X_train_resampled' in locals() or not 'y_train_resampled' in locals():
     print("ERROR: Resampled training data not found.")
     raise NameError("Resampled data missing.")


try:
    # Instantiate the classifier
    model_xgb_baseline = xgb.XGBClassifier(
        random_state=42,          # For reproducibility of internal randomness
        use_label_encoder=False,  # Avoids potential deprecation warnings
        eval_metric='logloss'     # Common metric, avoids potential warnings if not set
    )
    print("XGBoost Classifier instantiated.")
    print("Training XGBoost model on resampled data (takes time)...")


    # Train the model using the balanced (SMOTE'd) training data
    model_xgb_baseline.fit(X_train_resampled, y_train_resampled)


    print("Baseline model training complete.")
except Exception as e:
    print(f"ERROR during XGBoost training: {e}")
    raise


--- Step 1.5: Train Baseline XGBoost Model ---
XGBoost Classifier instantiated.
Training XGBoost model on resampled data (takes time)...


E:\AI Prep\Projects\Project 1\realtime-fraud-detection-api\p1env\Lib\site-packages\xgboost\training.py:183: UserWarning: [07:16:51] WARNING: C:\actions-runner\_work\xgboost\xgboost\src\learner.cc:738: 
Parameters: { "use_label_encoder" } are not used.

  bst.update(dtrain, iteration=i, fobj=obj)


Baseline model training complete.


--- Step 1.5: Train Baseline XGBoost Model ---: Indicates the start of this step.
XGBoost Classifier instantiated.: Confirms that the xgb.XGBClassifier(...) line executed correctly, creating the model object before training.
Training XGBoost model on resampled data (takes time)...: Shows that the model_xgb_baseline.fit(X_train_resampled, y_train_resampled) command started. This is where the actual learning happens, using the SMOTE-balanced data.
UserWarning: [...] Parameters: { "use_label_encoder" } are not used.:
What it means: XGBoost is telling you that even though you provided the parameter use_label_encoder=False when creating the classifier, this parameter is essentially ignored or not needed by the underlying training mechanism in newer versions, especially when you also specify eval_metric.
Why it happens: use_label_encoder was primarily relevant for older versions or specific internal label handling that is now deprecated or automatically handled differently. Setting it to False is the recommended practice to avoid potential future issues, but the library is just letting you know it didn't actively use that specific setting during this particular training run.
Is it a problem? No. This is a common warning and can be safely ignored. It does not mean the training failed or is incorrect.
Baseline model training complete.: This is the most important message! It confirms that the .fit() process finished successfully without crashing.

In [11]:
#Sub-Step 1.6: Evaluate Baseline Model
from sklearn.metrics import classification_report, recall_score, confusion_matrix
print("\n--- Step 1.6: Evaluate Baseline Model on Test Set ---")


# Ensure model and test data exist
if not 'model_xgb_baseline' in locals(): raise NameError("Baseline model not trained.")
if not 'X_test' in locals() or not 'y_test' in locals(): raise NameError("Test data not available.")


try:
    print("Making predictions on the original (imbalanced) test set...")
    # Use the trained model to predict on the unseen X_test
    y_pred_baseline = model_xgb_baseline.predict(X_test)


    print("\nConfusion Matrix (Baseline):")
    # Rows = Actual (0: Not Fraud, 1: Fraud)
    # Cols = Predicted (0: Not Fraud, 1: Fraud)
    cm = confusion_matrix(y_test, y_pred_baseline)
    print(cm)
    # You can manually interpret: TN=cm[0,0], FP=cm[0,1], FN=cm[1,0], TP=cm[1,1]


    print("\nClassification Report (Baseline):")
    print(classification_report(y_test, y_pred_baseline, target_names=['Not Fraud (0)', 'Fraud (1)'], digits=4))


    # Calculate recall specifically for the positive 'Fraud' class (label=1)
    recall_fraud = recall_score(y_test, y_pred_baseline, pos_label=1)
    print(f"\nRecall for Fraud Class (1): {recall_fraud:.4f}")


    # Check MVP goal
    MVP_RECALL_TARGET = 0.75 # Our defined target
    print(f"\nChecking against MVP Recall Target ({MVP_RECALL_TARGET})...")
    if recall_fraud >= MVP_RECALL_TARGET:
        print(f">>> Level 1 MVP Recall target MET! :) <<<")
    else:
        print(f">>> Level 1 MVP Recall target NOT MET. :( <<<")
        print(f"    (Further tuning in Level 2 or revisiting preprocessing might be needed)")
except Exception as e:
    print(f"ERROR during baseline model evaluation: {e}")
    raise


--- Step 1.6: Evaluate Baseline Model on Test Set ---
Making predictions on the original (imbalanced) test set...

Confusion Matrix (Baseline):
[[199141    752]
 [    13     94]]

Classification Report (Baseline):
               precision    recall  f1-score   support

Not Fraud (0)     0.9999    0.9962    0.9981    199893
    Fraud (1)     0.1111    0.8785    0.1973       107

     accuracy                         0.9962    200000
    macro avg     0.5555    0.9374    0.5977    200000
 weighted avg     0.9995    0.9962    0.9977    200000


Recall for Fraud Class (1): 0.8785

Checking against MVP Recall Target (0.75)...
>>> Level 1 MVP Recall target MET! :) <<<


!pip freeze > requirements.txt

--- LEVEL 2 (Refined Model) ---

In [12]:
#Sub-Step 2.1: Feature Engineering
import pandas as pd
import numpy as np
from imblearn.over_sampling import SMOTE # Re-import if in new notebook


print("\n--- Step 2.1: Feature Engineering (Example) ---")


# Ensure original train/test splits exist from Level 1
if not 'X_train' in locals() or not 'X_test' in locals() or not 'y_train' in locals():
     print("ERROR: Original train/test splits (X_train, X_test, y_train) not found.")
     raise NameError("Original splits missing.")


# Create copies to avoid modifying original dataframes
X_train_fe = X_train.copy()
X_test_fe = X_test.copy()
print(f"Original X_train shape: {X_train_fe.shape}")


# Example 1: Amount / Origin Balance Ratio
epsilon = 1e-6 # Small value to prevent division by zero
X_train_fe['amt_ratio_orig'] = (X_train_fe['amount'] / (X_train_fe['oldbalanceOrg'] + epsilon)).fillna(0)
X_test_fe['amt_ratio_orig'] = (X_test_fe['amount'] / (X_test_fe['oldbalanceOrg'] + epsilon)).fillna(0)
print("Added 'amt_ratio_orig' feature.")


# Example 2: Balance Discrepancy (Optional - can add complexity)
# X_train_fe['orig_bal_discrepancy'] = (X_train_fe['oldbalanceOrg'] - X_train_fe['newbalanceOrig']) - X_train_fe['amount']
# X_test_fe['orig_bal_discrepancy'] = (X_test_fe['oldbalanceOrg'] - X_test_fe['newbalanceOrig']) - X_test_fe['amount']
# print("Added 'orig_bal_discrepancy' feature.")


print(f"X_train_fe shape after FE: {X_train_fe.shape}")
print(f"X_test_fe shape after FE: {X_test_fe.shape}")
print("\nFirst 3 rows of X_train_fe with new feature(s):")
print(X_train_fe.head(3))


# --- Re-apply SMOTE on the NEW feature-engineered training data ---
# We need to balance the training set *after* adding features
print("\nRe-applying SMOTE to feature-engineered training data...")
try:
    smote_fe = SMOTE(random_state=50)
    # Use the feature-engineered training data (X_train_fe) and original y_train
    X_train_fe_resampled, y_train_fe_resampled = smote_fe.fit_resample(X_train_fe, y_train)
    print("SMOTE application complete on feature-engineered data.")
    print(f"X_train_fe_resampled shape: {X_train_fe_resampled.shape}")
    print(f"y_train_fe_resampled shape: {y_train_fe_resampled.shape}")
    print(f"\nResampled class distribution:\n{y_train_fe_resampled.value_counts(normalize=True)}")
except Exception as e:
    print(f"ERROR during SMOTE on feature-engineered data: {e}")
    raise


--- Step 2.1: Feature Engineering (Example) ---
Original X_train shape: (800000, 10)
Added 'amt_ratio_orig' feature.
X_train_fe shape after FE: (800000, 11)
X_test_fe shape after FE: (200000, 11)

First 3 rows of X_train_fe with new feature(s):
        step    amount  oldbalanceOrg  newbalanceOrig  oldbalanceDest  \
22807      8  47249.07            0.0            0.00      1040812.95   
18282      8   2551.07        25280.0        22728.93            0.00   
916408    43  72301.09        72908.0          606.91            0.00   

        newbalanceDest  type_CASH_OUT  type_DEBIT  type_PAYMENT  \
22807       1930535.01           True       False         False   
18282             0.00          False       False          True   
916408       123056.71           True       False         False   

        type_TRANSFER  amt_ratio_orig  
22807           False    4.724907e+10  
18282           False    1.009126e-01  
916408          False    9.916757e-01  

Re-applying SMOTE to feature-en

In [13]:
#Sub-Step 2.2: Build Scikit-learn Pipeline
from sklearn.pipeline import Pipeline
# from sklearn.preprocessing import StandardScaler # Optional: if you wanted to add scaling
import xgboost as xgb # Re-import if needed


print("\n--- Step 2.2: Build Scikit-learn Pipeline ---")


# Define the steps for the pipeline
pipeline_steps = [
    # Optional Step: Scaling (StandardScaler makes features have zero mean, unit variance)
    # Though XGBoost is not sensitive to scaling, it's good practice if comparing models
    # ('scaler', StandardScaler()),


    # Final Step: The XGBoost Classifier
    # Use the step name 'xgboost' to match parameter names later in tuning
    ('xgboost', xgb.XGBClassifier(random_state=50, use_label_encoder=False, eval_metric='logloss'))
]


# Create the Pipeline object
pipeline_obj = Pipeline(steps=pipeline_steps)


print("Pipeline object created successfully:")
print(pipeline_obj)


--- Step 2.2: Build Scikit-learn Pipeline ---
Pipeline object created successfully:
Pipeline(steps=[('xgboost',
                 XGBClassifier(base_score=None, booster=None, callbacks=None,
                               colsample_bylevel=None, colsample_bynode=None,
                               colsample_bytree=None, device=None,
                               early_stopping_rounds=None,
                               enable_categorical=False, eval_metric='logloss',
                               feature_types=None, feature_weights=None,
                               gamma=None, grow_policy=None,
                               importance_type=None,
                               interaction_constraints=None, learning_rate=None,
                               max_bin=None, max_cat_threshold=None,
                               max_cat_to_onehot=None, max_delta_step=None,
                               max_depth=None, max_leaves=None,
                               min_child_weight=

In [14]:
!pip install ray[tune] scikit-optimize hyperopt

In [15]:
#Sub-Step 2.3: Hyperparameter Tuning with Ray Tune
from ray.tune.integration.sklearn import TuneSearchCV
from scipy.stats import randint, uniform
import ray # Import ray if needed for initialization (sometimes automatic)


# Optional: Initialize Ray explicitly (might help manage resources)
# if not ray.is_initialized():
#    ray.init(ignore_reinit_error=True, num_cpus=4) # Adjust num_cpus if needed


print("\n--- Step 2.3: Hyperparameter Tuning with Ray Tune ---")


# Define parameter search space for XGBoost within the pipeline
# Prefix parameters with step name ('xgboost__')
param_distributions = {
    'xgboost__n_estimators': randint(150, 500),        # Number of trees
    'xgboost__max_depth': randint(4, 12),             # Max depth of trees
    'xgboost__learning_rate': uniform(0.01, 0.2),    # Step size shrinkage (loc=0.01, scale=0.2 -> range 0.01 to 0.21)
    'xgboost__subsample': uniform(0.6, 0.4),         # Fraction of samples used per tree (range 0.6 to 1.0)
    'xgboost__colsample_bytree': uniform(0.6, 0.4),  # Fraction of features used per tree (range 0.6 to 1.0)
    # Add other XGBoost params if desired, e.g., gamma, reg_alpha, reg_lambda
}


# Setup TuneSearchCV
tune_search = TuneSearchCV(
    pipeline_obj,              # The pipeline with (optional scaler and) XGBoost
    param_distributions,
    n_trials=15,               # Increase trials for better search (e.g., 15-30), but takes longer
    scoring='recall',          # Optimize for recall of the positive class (Fraud=1)
    cv=3,                      # 3-fold cross-validation
    random_state=50,
    n_jobs=-1,                 # Use available cores
    verbose=2                  # Show more output during tuning
)

print("Starting Ray Tune search (THIS CAN TAKE SIGNIFICANT TIME)...")
# Fit TuneSearchCV on the FEATURE-ENGINEERED, SMOTE-RESAMPLED training data
# Ensure these variables exist from Sub-Step 2.1
if not 'X_train_fe_resampled' in locals() or not 'y_train_fe_resampled' in locals():
     print("ERROR: Feature-engineered resampled training data not found.")
     raise NameError("Feature-engineered resampled data missing.")


try:
    tune_search.fit(X_train_fe_resampled, y_train_fe_resampled) # Fit on the balanced, feature-engineered training set!


    print("\nRay Tune search finished.")
    print(f"Best parameters found: {tune_search.best_params_}")
    print(f"Best cross-validation recall score during search: {tune_search.best_score_:.4f}")


    # Access the best pipeline refitted on the whole training data
    best_pipeline = tune_search.best_estimator_


except Exception as e:
    print(f"ERROR during Ray Tune search: {e}")
    # Check Ray logs if available for more details
    raise
# Optional: Shutdown Ray if initialized manually
# ray.shutdown()


ModuleNotFoundError: No module named 'ray.tune.integration.sklearn'

In [ ]:
#ray error checking 1
import sys
print("Python Executable:", sys.executable)

In [ ]:
#ray error checking 1
import importlib
try:
    ray_spec = importlib.util.find_spec("ray")
    if ray_spec:
        print("Found 'ray' installation at:", ray_spec.origin)
        # Try importing base ray and tune
        import ray
        print("Successfully imported base 'ray'. Version:", ray.__version__)
        import ray.tune
        print("Successfully imported 'ray.tune'.")
        # Check if the specific module path exists conceptually
        if hasattr(ray.tune, 'integration') and hasattr(ray.tune.integration, 'sklearn'):
            print("'ray.tune.integration.sklearn' path seems accessible.")
            # Try the final import again
            from ray.tune.integration.sklearn import TuneSearchCV
            print("SUCCESS: Imported TuneSearchCV!")
        else:
             print("ERROR: 'ray.tune.integration.sklearn' structure not found in this ray version.")

    else:
        print("ERROR: 'ray' package not found in this environment.")
except ImportError as e:
    print(f"ERROR during import: {e}")
except Exception as e:
    print(f"An unexpected error occurred: {e}")

In [ ]:
#ray checking 3
import ray.tune
import inspect

print("--- Inspecting ray.tune submodules ---")
found_locations = []
possible_locations = ['sklearn', 'integration', 'integration.sklearn'] # Things we've tried or expect

for name, obj in inspect.getmembers(ray.tune):
    if inspect.ismodule(obj):
        print(f"Found submodule: ray.tune.{name}")
        # Check if it contains TuneSearchCV or is one of our target paths
        if name in possible_locations:
            found_locations.append(f"ray.tune.{name}")
        try:
            if hasattr(obj, 'TuneSearchCV'):
                 print(f"  >>> Found TuneSearchCV in: ray.tune.{name}")
                 found_locations.append(f"ray.tune.{name}.TuneSearchCV")
            # Check one level deeper for integration.sklearn
            if name == 'integration' and hasattr(obj, 'sklearn'):
                 if hasattr(obj.sklearn, 'TuneSearchCV'):
                      print(f"  >>> Found TuneSearchCV in: ray.tune.integration.sklearn")
                      found_locations.append(f"ray.tune.integration.sklearn.TuneSearchCV")
        except Exception as e:
            print(f"  (Could not inspect submodule {name} fully: {e})")


if not found_locations:
    print("\nERROR: Could not find TuneSearchCV in expected ray.tune submodules.")
    print("This suggests the 'ray[tune]' installation might be incomplete or corrupted,")
    print("or TuneSearchCV is located elsewhere in Ray v2.45.0.")
else:
    print("\nPossible locations containing TuneSearchCV or relevant modules:")
    for loc in set(found_locations): # Use set to remove duplicates
         print(f"- {loc}")

In [ ]:
#ray checking 4
from ray.tune.integration.sklearn import TuneSearchCV
print("Successfully imported TuneSearchCV!")

In [16]:
# --- MODIFIED Sub-Step 2.3: Hyperparameter Tuning with RandomizedSearchCV ---
# from ray.tune.integration.sklearn import TuneSearchCV # REMOVED RAY IMPORT
from sklearn.model_selection import RandomizedSearchCV # USE SKLEARN'S TOOL
from scipy.stats import randint, uniform
# import ray # No longer needed for this step

print("\n--- Step 2.3: Hyperparameter Tuning with RandomizedSearchCV ---")

# --- Ensure pipeline_obj exists from Step 2.2 ---
if not 'pipeline_obj' in locals():
     print("ERROR: 'pipeline_obj' (Pipeline object from Step 2.2) not found.")
     raise NameError("'pipeline_obj' is missing.")

# --- Define parameter search space (SAME AS BEFORE) ---
# Prefix parameters with step name ('xgboost__')
param_distributions = {
    'xgboost__n_estimators': randint(150, 500),        # Number of trees
    'xgboost__max_depth': randint(4, 12),             # Max depth of trees
    'xgboost__learning_rate': uniform(0.01, 0.2),    # Step size shrinkage (range 0.01 to 0.21)
    'xgboost__subsample': uniform(0.6, 0.4),         # Fraction of samples used per tree (range 0.6 to 1.0)
    'xgboost__colsample_bytree': uniform(0.6, 0.4),  # Fraction of features used per tree (range 0.6 to 1.0)
}

# --- Setup RandomizedSearchCV (INSTEAD OF TuneSearchCV) ---
n_iterations = 15 # Number of parameter settings that are sampled. n_iter trades off runtime vs quality of the solution.
print(f"Setting up RandomizedSearchCV with {n_iterations} iterations...")

random_search = RandomizedSearchCV(
    pipeline_obj,              # The pipeline object
    param_distributions=param_distributions, # The distributions to sample from
    n_iter=n_iterations,       # Number of combinations to try (like n_trials)
    scoring='recall',          # Metric to optimize for (use 'recall_macro' or specific scorer if needed)
    cv=3,                      # 3-fold cross-validation
    random_state=50,           # For reproducibility (use 42 if preferred from plan)
    n_jobs=-1,                 # Use all available CPU cores for CV folds
    verbose=2                  # Show progress
)

print("Starting RandomizedSearch search (THIS CAN TAKE SIGNIFICANT TIME)...")
# Ensure feature-engineered, resampled training data exists from Step 2.1
if not 'X_train_fe_resampled' in locals() or not 'y_train_fe_resampled' in locals():
     print("ERROR: Feature-engineered resampled training data not found.")
     raise NameError("Feature-engineered resampled data missing.")

try:
    # Fit RandomizedSearchCV on the FEATURE-ENGINEERED, SMOTE-RESAMPLED training data
    random_search.fit(X_train_fe_resampled, y_train_fe_resampled)

    print("\nRandomizedSearch search finished.")
    print(f"Best parameters found: {random_search.best_params_}")
    print(f"Best cross-validation recall score during search: {random_search.best_score_:.4f}")

    # Access the best pipeline refitted on the whole training data
    best_pipeline = random_search.best_estimator_

except Exception as e:
    print(f"ERROR during RandomizedSearch search: {e}")
    raise


--- Step 2.3: Hyperparameter Tuning with RandomizedSearchCV ---
Setting up RandomizedSearchCV with 15 iterations...
Starting RandomizedSearch search (THIS CAN TAKE SIGNIFICANT TIME)...
Fitting 3 folds for each of 15 candidates, totalling 45 fits


E:\AI Prep\Projects\Project 1\realtime-fraud-detection-api\p1env\Lib\site-packages\xgboost\training.py:183: UserWarning: [07:27:10] WARNING: C:\actions-runner\_work\xgboost\xgboost\src\learner.cc:738: 
Parameters: { "use_label_encoder" } are not used.

  bst.update(dtrain, iteration=i, fobj=obj)



RandomizedSearch search finished.
Best parameters found: {'xgboost__colsample_bytree': np.float64(0.9157080755356058), 'xgboost__learning_rate': np.float64(0.08288386811018128), 'xgboost__max_depth': 8, 'xgboost__n_estimators': 472, 'xgboost__subsample': np.float64(0.8673350275106719)}
Best cross-validation recall score during search: 1.0000


In [18]:
#Sub-Step 2.4: Evaluate Tuned Model
from sklearn.metrics import classification_report, recall_score, confusion_matrix
print("\n--- Step 2.4: Evaluate Tuned Model on Test Set ---")


# Ensure best_pipeline and X_test_fe exist
if not 'best_pipeline' in locals():
     print("ERROR: 'best_pipeline' not found. Tuning may have failed.")
     raise NameError("'best_pipeline' not defined.")
if not 'X_test_fe' in locals() or not 'y_test' in locals():
     print("ERROR: Feature-engineered test data (X_test_fe) or y_test not found.")
     raise NameError("Test data missing.")


try:
    print("Making predictions with the tuned pipeline on the feature-engineered test set...")
    # Predict using the best pipeline on X_test_fe
    y_pred_tuned = best_pipeline.predict(X_test_fe)


    print("\nConfusion Matrix (Tuned):")
    print(confusion_matrix(y_test, y_pred_tuned)) # Compare against original y_test


    print("\nClassification Report (Tuned):")
    print(classification_report(y_test, y_pred_tuned, target_names=['Not Fraud (0)', 'Fraud (1)'], digits=4))


    # Calculate recall for the positive 'Fraud' class
    recall_fraud_tuned = recall_score(y_test, y_pred_tuned, pos_label=1)
    print(f"\nRecall for Fraud Class (1) - Tuned: {recall_fraud_tuned:.4f}")


    # Optional: Compare with baseline recall if stored from L1
    if 'recall_fraud' in locals():
          print(f"(Recall Baseline was: {recall_fraud:.4f})")
          if recall_fraud_tuned > recall_fraud: print(">>> Recall IMPROVED vs Baseline.")
          else: print(">>> Recall did NOT improve vs Baseline.")


except Exception as e:
    print(f"ERROR during tuned model evaluation: {e}")
    raise


--- Step 2.4: Evaluate Tuned Model on Test Set ---
Making predictions with the tuned pipeline on the feature-engineered test set...

Confusion Matrix (Tuned):
[[199683    210]
 [    13     94]]

Classification Report (Tuned):
               precision    recall  f1-score   support

Not Fraud (0)     0.9999    0.9989    0.9994    199893
    Fraud (1)     0.3092    0.8785    0.4574       107

     accuracy                         0.9989    200000
    macro avg     0.6546    0.9387    0.7284    200000
 weighted avg     0.9996    0.9989    0.9992    200000


Recall for Fraud Class (1) - Tuned: 0.8785
(Recall Baseline was: 0.8785)
>>> Recall did NOT improve vs Baseline.


In [19]:
#Sub-Step 2.5: Save Artifacts & Commit Level 2 Completion

import joblib # Or use pickle
print("\n--- Step 2.5: Save Best Pipeline & Commit ---")


pipeline_filename = "best_fraud_pipeline.joblib"
if 'best_pipeline' in locals():
     try:
         joblib.dump(best_pipeline, pipeline_filename)
         print(f"Saved best pipeline object to '{pipeline_filename}'")
     except Exception as e:
         print(f"ERROR saving pipeline: {e}")
else:
     print("WARNING: 'best_pipeline' object not found, cannot save.")


--- Step 2.5: Save Best Pipeline & Commit ---
Saved best pipeline object to 'best_fraud_pipeline.joblib'


In [20]:
!pip freeze > requirements.txt